### PR TITLE
executor: unify executor concurrency

### LINKED ISSUE #15428: UCP: Control the concurrency of all types of executors in one session variable
## Description

Currently, TiDB has several variables to control each type of executor,

See https://github.com/pingcap/tidb/blob/45d83d1adb3eedc83bf5f59f21e6666dfcf8696f/sessionctx/variable/session.go#L1282-L1309

And each of them has its own default value.
I propose to add one session variable named like`ExecutorConcurrency` to control the concurrency of all types of executors and make the default value as the number of the CPU cores.

For compatibility of the old variables, take `HashJoinConcurrency` as an example,
1) The default value of `ExecutorConcurrency` is  the number of the CPU cores.
2) Remove the default value of `HashJoinConcurrency`. 
3) If `HashJoinConcurrency` is unset, the concurrency of Hash Join is applied by `ExecutorConcurrency`. If `select @tidb_hash_join_concurrency;` is queried, the value of `ExecutorConcurrency` is outputted.
4) If `HashJoinConcurrency` is set specifically,  the concurrency of Hash Join is applied by `HashJoinConcurrency`. If `select @tidb_hash_join_concurrency;` is queried, the value of `HashJoinConcurrency` is outputted.

## Score

* 300

## Mentor(s)

* @SunRunAway 

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills

* Golang

## Learning Materials

* Chinese: [TiDB 精选技术讲解文章](https://github.com/pingcap/presentations/blob/master/hackathon-2019/reference-document-of-hackathon-2019.md)
* English: [Awesome-Database-Learning](https://github.com/pingcap/awesome-database-learning)

### PR DESCRIPTION
**Pull request by niedhui**
UCP #15428 

### What problem does this PR solve?

Issue Number: part of  #15428

Problem Summary:

### What is changed and how it works?

What's Changed:

Deprecated existing executor-concurrency related variables,  use a single `ExecutorConcurreny` to control the concurrency of all types of executors.

How it Works:
1. Introduce a system variable  `ExecutorConcurrency`  to control the concurrency of all types of executors.
2. All deprecated fields of `sessionctx/Concurrency` were hidden, and use the exported functions to get the value
3. The functions first try to read the deprecated-field for compatibility, if the value was -1 (which means unset),  the value of `ExecutorConcurrency` was returned.

**Updates** 

We have **nine** executor-concurrency related variables:

- tidb_index_lookup_concurrency
- tidb_index_lookup_join_concurrency
- tidb_hashagg_partial_concurrency
- tidb_hashagg_final_concurrency
- tidb_window_concurrency
- tidb_projection_concurrency
- tidb_hash_join_concurrency
- tidb_distsql_scan_concurrency
- tidb_index_serial_scan_concurrency

I planed to migrate these variables in the following steps:

1. In this PR, I only migrate the first **6** variables to adopt `ExecutorConcurrency`. Because the original default value of these variables is all **4**, by making the default value of `ExecutorConcurrency` to **4** (not the number of cups), we would not break too many tests
2. I will migrate `tidb_hash_join_concurrency` and `tidb_distsql_scan_concurrency` in the following PR, the original default value of these two is **5** and **15**, we may discuss the default fixed-value of `ExecutorConcurreny` in the PR, **4** or **15**, and this will break many tests,  in `TiDB`, `mysql-tester` and `copr-test`, I would make changes on these repos by setting the fixed value of the concurrency variables.
3. As we already knew, some tests are in strict mode, we may add the deprecated warning in a  separate PR as the last step.
4. `tidb_index_serial_scan_concurrency` will be kept since it's used by analyze-table as @SunRunAway suggested.

Known issues about tests currently :
We should only use `set @@session.tidb_executor_concurrenly=` to control the concurrency of executor, the execution plan may not be the same as it is, which may lead to lots of changes for currency tests.

### Check List 

- [x] All executors use `ExecutorConcurrency` 
- [x] Bootstrap from old version
- [x] Show variable and values in `mysql.global_variables`
- [x] fix test

Tests

- Unit test
- Integration test

### Release note 
Deprecate `tidb_index_lookup_concurrency`, `tidb_index_lookup_join_concurrency`, ``, `tidb_hashagg_partial_concurrency`, `tidb_hashagg_final_concurrency`, `tidb_window_concurrency`, `tidb_projection_concurrency` and `` use the new `tidb_executor_concurrency` instead.

### CODE REVIEW

**Comment by niedhui** [comment]
@SunRunAway friendly ping 😀

File: sessionctx/variable/tidb_vars.go [hunk 1]
```diff
@@ -509,4 +507,5 @@ var (
 	ExpensiveQueryTimeThreshold    uint64 = DefTiDBExpensiveQueryTimeThreshold
 	MinExpensiveQueryTimeThreshold uint64 = 10 //10s
 	CapturePlanBaseline                   = serverGlobalVariable{globalVal: "0"}
+	DefExecutorConcurrency                = runtime.NumCPU()
```
Comment (SunRunAway) [inline sessionctx/variable/tidb_vars.go:510]: Is it suitable in Kubernetes environment? @Yisaer

**Comment by niedhui** [comment]
🤣 Welcome back, mentor.  I'll update this tomorrow

File: sessionctx/variable/tidb_vars.go (hunk 1 above)
Comment (Yisaer) [reply sessionctx/variable/tidb_vars.go:510]: I'm afraid not. `runtime.NumCPU()` would get the Host CPU Core nums which is not kubernetes-friendly.

**Comment by niedhui** [comment]
@SunRunAway 

>  but only deprecate and hide to new users for tidb v4.0.

Then will the user still be able to change these deprecated variables using SQL? like 
```
set @@session.tidb_hashagg_partial_concurrency = 1
```

**Comment by SunRunAway** [comment]
> @SunRunAway
> 
> > but only deprecate and hide to new users for tidb v4.0.
> 
> Then will the user still be able to change these deprecated variables using SQL? like
> 
> ```
> set @@session.tidb_hashagg_partial_concurrency = 1
> ```

Yes. And the code is like,

if `tidb_hashagg_partial_concurrency` is not `NULL`:
    concurrency = tidb_hashagg_partial_concurrency
else 
    concurrency = tidb_executor_concurrency

File: sessionctx/variable/tidb_vars.go (hunk 1 above)
Comment (SunRunAway) [reply sessionctx/variable/tidb_vars.go:510]: If we cannot get a quick fix for the container environment, we can leave it as an issue and try to find a way in future.

File: sessionctx/variable/varsutil_test.go [hunk 2]
```diff
@@ -60,16 +60,17 @@ func (s *testVarsutilSuite) TestNewSessionVars(c *C) {
 
 	c.Assert(vars.IndexJoinBatchSize, Equals, DefIndexJoinBatchSize)
 	c.Assert(vars.IndexLookupSize, Equals, DefIndexLookupSize)
-	c.Assert(vars.IndexLookupConcurrency, Equals, DefIndexLookupConcurrency)
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply sessionctx/variable/varsutil_test.go:63]: If you wish to keep `vars.IndexLookupConcurrency` as `vars.indexLookupConcurrency` as an unexported type, an integration test containing `select @@tidb_index_lookup_concurrency;` can be added into this file.

File: sessionctx/variable/session.go [hunk 3]
```diff
@@ -1384,34 +1387,156 @@ type TableDelta struct {
...
-	HashAggPartialConcurrency int
+	hashAggPartialConcurrency int
 
 	// HashAggFinalConcurrency is the number of concurrent hash aggregation final worker.
-	HashAggFinalConcurrency int
+	hashAggFinalConcurrency int
 
 	// WindowConcurrency is the number of concurrent window worker.
-	WindowConcurrency int
+	windowConcurrency int
```
Comment (SunRunAway) [inline sessionctx/variable/session.go:1416]: The comments for the variables should be updated.

File: sessionctx/variable/tidb_vars.go [hunk 4]
```diff
@@ -396,6 +397,9 @@ const (
 
 	// TiDBEnableCollectExecutionInfo indicates that whether execution info is collected.
 	TiDBEnableCollectExecutionInfo = "tidb_enable_collect_execution_info"
+
+	// DefExecutorConcurrency is used for controlling the concurrency of all types of executors.
+	TiDBExecutorConcurrency = "tidb_executor_concurrency"
```
Comment (SunRunAway) [inline sessionctx/variable/tidb_vars.go:402]: The old and deprecated variables should be documented.

**Comment by SunRunAway** [comment]
And best to update the PR description for your newest code and add a release note for the check_release_note bot.

**Comment by SunRunAway** [comment]
@niedhui A friendly ping.

**Comment by niedhui** [comment]
@SunRunAway 
I may push some codes later today. And two more things to confirm:
1. Since the deprecated variables like `tidb_hash_join_concurrency` are unset by default, the value will be stored in table `mysql. global_variables`, if the users query this table,  they will see the value is -1, although the executors are using calculated value, is it OK?  Or do you have any suggestions?

2. Change the default values of these variables seems to break lots of tests, especially `explain tests` under `cmd/explaintest`.  I need to set a fixed value of `ExecutorConcurreny` for each test and regenerate the result, right?

**Comment by SunRunAway** [comment]
> @SunRunAway
> I may push some codes later today. And two more things to confirm:
> 
> 1. Since the deprecated variables like `tidb_hash_join_concurrency` are unset by default, the value will be stored in table `mysql. global_variables`, if the users query this table,  they will see the value is -1, although the executors are using calculated value, is it OK?  Or do you have any suggestions?
> 2. Change the default values of these variables seems to break lots of tests, especially `explain tests` under `cmd/explaintest`.  I need to set a fixed value of `ExecutorConcurreny` for each test and regenerate the result, right?

> they will see the value is -1, although the executors are using calculated value, is it OK?

How about using 0 instead of -1?

**Comment by SunRunAway** [comment]
> they will see the value is -1, although the executors are using calculated value, is it OK?

How about using 0 instead of -1?

> I need to set a fixed value of ExecutorConcurreny for each test and regenerate the result, right?

Yes, that's a suitable way.

**Comment by niedhui** [comment]
@SunRunAway `0` is a valid value for `TiDBProjectionConcurrency` https://github.com/pingcap/tidb/blob/9e16c59ef0e6b66f861a714cdae50f0cf1a6f447/sessionctx/variable/varsutil.go#L569

File: executor/set_test.go [hunk 5]
```diff
@@ -568,13 +568,13 @@ func (s *testSuite5) TestValidateSetVar(c *C) {
...
 
-	_, err = tk.Exec("set global tidb_distsql_scan_concurrency=-1;")
+	_, err = tk.Exec("set global tidb_distsql_scan_concurrency=-2;")
 	c.Assert(terror.ErrorEqual(err, variable.ErrWrongValueForVar), IsTrue, Commentf("err %v", err))
 
 	_, err = tk.Exec("set @@tidb_distsql_scan_concurrency='fff';")
 	c.Assert(terror.ErrorEqual(err, variable.ErrWrongTypeForVar), IsTrue, Commentf("err %v", err))
 
-	_, err = tk.Exec("set @@tidb_distsql_scan_concurrency=-1;")
+	_, err = tk.Exec("set @@tidb_distsql_scan_concurrency=-2;")
```
Comment (SunRunAway) [inline executor/set_test.go:577]: This change is not necessary.

File: executor/builder.go [hunk 6]
```diff
@@ -2199,9 +2199,9 @@ func (b *executorBuilder) buildIndexNestedLoopHashJoin(v *plannercore.PhysicalIn
 		IndexLookUpJoin: *e,
 		keepOuterOrder:  v.KeepOuterOrder,
 	}
-	concurrency := e.ctx.GetSessionVars().IndexLookupJoinConcurrency
-	idxHash.joiners = make([]joiner, concurrency)
-	for i := 0; i < concurrency; i++ {
+	currency := e.ctx.GetSessionVars().IndexLookupJoinConcurrency()
+	idxHash.joiners = make([]joiner, currency)
+	for i := 0; i < currency; i++ {
```
[the remark this answers is not in the document]
Comment (SunRunAway) [reply executor/builder.go:2227]: Please resolve this comment.

File: sessionctx/variable/varsutil_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (SunRunAway) [reply sessionctx/variable/varsutil_test.go:63]: > an integration test containing `select @@tidb_index_lookup_concurrency;` can be added into this file.

Would you like to add an integration test here or in executor/set_test.go?

File: sessionctx/variable/session.go [hunk 7]
```diff
@@ -1426,34 +1429,167 @@ type TableDelta struct {
...
+	hashAggPartialConcurrency *int
+
+	// hashAggFinalConcurrency is the number of concurrent hash aggregation final worker.
+	hashAggFinalConcurrency *int
+
+	// windowConcurrency is the number of concurrent window worker.
+	windowConcurrency *int
+
+	// indexSerialScanConcurrency is the number of concurrent index serial scan worker.
+	indexSerialScanConcurrency int
```
Comment (SunRunAway) [inline sessionctx/variable/session.go:1474]: These should be documented as deprecated.

File: sessionctx/variable/session.go [hunk 8]
```diff
@@ -1134,29 +1135,31 @@ func (s *SessionVars) SetSystemVar(name string, val string) error {
 	case TiDBOptConcurrencyFactor:
 		s.ConcurrencyFactor = tidbOptFloat64(val, DefOptConcurrencyFactor)
 	case TiDBIndexLookupConcurrency:
-		s.IndexLookupConcurrency = tidbOptPositiveInt32(val, DefIndexLookupConcurrency)
+		s.indexLookupConcurrency = tidbOptPositiveInt32Ptr(val)
```
Comment (SunRunAway) [inline sessionctx/variable/session.go:1138]: Better to append a warning message like https://github.com/pingcap/tidb/blob/622d40b5f28afd92a6d317fb8daf714afbdf545f/sessionctx/variable/session.go#L1187

File: sessionctx/variable/session.go [hunk 9]
```diff
@@ -1426,34 +1429,167 @@ type TableDelta struct {
...
+	// indexSerialScanConcurrency is the number of concurrent index serial scan worker.
+	indexSerialScanConcurrency int
+
+	// ExecutorConcurrency is the number of concurrent worker for all executors.
+	ExecutorConcurrency int
+}
+
+// SetIndexLookupConcurrency set the number of concurrent index lookup worker.
+func (c *Concurrency) SetIndexLookupConcurrency(n int) {
+	c.indexLookupConcurrency = &n
```
Comment (zz-jason) [inline sessionctx/variable/session.go:1482]: why using pointers to store basic int value? It points to a parameter address that is allocated when the function called.

**Comment by niedhui** [comment]
Hi @SunRunAway 

In the latest commit, I tried to use `NULL` as the default value of the deprecated concurrency variable, and it seems not a good solution(IMHO).

**First**, although the `VARIABLE_VALUE` of  `GLOBAL_VARIABLES` table accepts NULL
```
mysql> desc global_variables;
+----------------+---------------+------+------+---------+-------+
| Field          | Type          | Null | Key  | Default | Extra |
+----------------+---------------+------+------+---------+-------+
| VARIABLE_NAME  | varchar(64)   | NO   | PRI  | NULL    |       |
| VARIABLE_VALUE | varchar(1024) | YES  |      | NULL    |       |
+----------------+---------------+------+------+---------+-------+
2 rows in set (0.00 sec)
```
but  in `SysVar`, the value is always a string
https://github.com/pingcap/tidb/blob/f87e2b17102c07f3298a89289ae4d9c0e8e253a6/sessionctx/variable/sysvar.go#L41-L50
and the system variables stored in `Session`
https://github.com/pingcap/tidb/blob/f87e2b17102c07f3298a89289ae4d9c0e8e253a6/sessionctx/variable/session.go#L312-L323

So we need to treat the empty string as `NULL` for these integer types. 
when using `select @@tidb_index_lookup_concurrency`, we need to transfer the empty string to string `NULL`, 
```
	if _, ok := variable.DeprecatedConcurrencyVariables[name]; ok && val == variable.NullString {
		val = "NULL"
	}
```
The integer fields of `Concurrency` need to be a pointer or something like `sql.NullInt64`, or add a bit flag field to have the `NULL` meaning.

All these logic looks like patch codes.

**Secondly**, in MySQL, they using -1 to signifies autosizing value of integer type, 
https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_host_cache_size
if we using -1, the break-change is for `TidbProjectionConcurrency`, it accepts negative values before.

Any suggestion?

File: sessionctx/variable/session.go (hunk 8 above)
Comment (niedhui) [reply sessionctx/variable/session.go]: done

File: sessionctx/variable/session.go (hunk 7 above)
Comment (niedhui) [reply sessionctx/variable/session.go]: done

File: sessionctx/variable/varsutil_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (niedhui) [reply sessionctx/variable/varsutil_test.go:63]: done

File: executor/builder.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (niedhui) [reply executor/builder.go:2224]: done

File: sessionctx/variable/session.go [hunk 10]
```diff
@@ -737,15 +737,16 @@ func NewSessionVars() *SessionVars {
 	}
 	vars.KVVars = kv.NewVariables(&vars.Killed)
 	vars.Concurrency = Concurrency{
-		IndexLookupConcurrency:     DefIndexLookupConcurrency,
```
Comment (SunRunAway) [inline sessionctx/variable/session.go:740]: We can set `DefIndexLookupConcurrency` as `ConcurrencyUnset`

File: sessionctx/variable/session.go (hunk 9 above)
Comment (niedhui) [reply sessionctx/variable/session.go]: Had rollback to using int directly.

File: sessionctx/variable/session.go (hunk 10 above)
Comment (niedhui) [reply sessionctx/variable/session.go:740]: Updated. PTAL

**Comment by niedhui** [comment]
Since this PR changed `DefDistSQLScanConcurrency` from `15` to `-1`,  which is used by `br`, I'll open another PR in `br`, change it to using the new `ExecutorConcurrency`

File: sessionctx/variable/session.go [hunk 11]
```diff
@@ -1432,34 +1443,163 @@ type TableDelta struct {
...
+	// distSQLScanConcurrency is deprecated, use ExecutorConcurrency instead.
+	distSQLScanConcurrency int
+
+	// hashJoinConcurrency is the number of concurrent hash join outer worker.
+	// hashJoinConcurrency is deprecated, use ExecutorConcurrency instead.
+	hashJoinConcurrency int
+
+	// projectionConcurrency is the number of concurrent projection worker.
+	// projectionConcurrency is deprecated, use ExecutorConcurrency instead.
+	projectionConcurrency int64
```
Comment (zz-jason) [inline sessionctx/variable/session.go:1469]:
```suggestion
	projectionConcurrency int
```

**Comment by zz-jason** [comment]
```
[2020-06-16T13:07:04.798Z] time="2020-06-16T21:07:04+08:00" level=warning msg="Create new db&{0 {root:@tcp(127.0.0.1:4001)/test?strict=true&time_zone=%27Asia%2FShanghai%27 0x10a1110} 0 {0 0} [0xc000134280] map[] 0 1 0xc00002c2a0 0xc0000ca840 false map[0xc000134280:map[0xc000134280:true]] map[] 0 0 0 <nil> 0 0 0 0x48c430}"

[2020-06-16T13:07:04.798Z] time="2020-06-16T21:07:04+08:00" level=fatal msg="Executing \"SET @@tidb_hash_join_concurrency=2\" err[Warning 1287: 'tidb_hash_join_concurrency' is deprecated and will be removed in a future release. Please use tidb_executor_concurrency instead]"
```

**Comment by niedhui** [comment]
> ```
> [2020-06-16T13:07:04.798Z] time="2020-06-16T21:07:04+08:00" level=warning msg="Create new db&{0 {root:@tcp(127.0.0.1:4001)/test?strict=true&time_zone=%27Asia%2FShanghai%27 0x10a1110} 0 {0 0} [0xc000134280] map[] 0 1 0xc00002c2a0 0xc0000ca840 false map[0xc000134280:map[0xc000134280:true]] map[] 0 0 0 <nil> 0 0 0 0x48c430}"
> 
> [2020-06-16T13:07:04.798Z] time="2020-06-16T21:07:04+08:00" level=fatal msg="Executing \"SET @@tidb_hash_join_concurrency=2\" err[Warning 1287: 'tidb_hash_join_concurrency' is deprecated and will be removed in a future release. Please use tidb_executor_concurrency instead]"
> ```

Seems in `strict mode`, the warning was treated as error,  should I remove the warning temporarily ？

**Comment by niedhui** [comment]
@sre-bot /run-all-tests

**Comment by niedhui** [comment]
@sre-bot /run-integration-br-test br=pr/362

**Comment by niedhui** [comment]
@sre-bot /run-unit-test

**Comment by SunRunAway** [comment]
image

@kennytm It seems that the BR command-line binary does not use the version in this pull request as dependency.

**Comment by kennytm** [comment]
@SunRunAway We could modify the BR CI to add a replace the `github.com/pingcap/tidb` dependency by the specific TiDB commit, but the TiDB CI must be modified to *tell* the BR CI which TiDB commit is used. Someone from SRE may chime in?

Also, even if this trick allows the `integration-br-test` test to pass, BR itself would still not be fixed until a manual update.

**Comment by niedhui** [comment]
Hi @zz-jason, I add an `Updates` section in the PR description. TLDR, in order to not break too many tests, I will only adopt some variables to `ExecutorConcurrency` and will open another PR to adopt the left. 
PTAL, thanks.

**Pull request merged**